Core/Src/main.c

```diff
@@ -1,181 +1,223 @@
+  */
+/* USER CODE END Header */
+/* Includes ------------------------------------------------------------------*/
+#include "main.h"
+#include "spi.h"
+#include "tim.h"
+#include "usart.h"
+#include "gpio.h"
+
+/* Private includes ----------------------------------------------------------*/
+/* USER CODE BEGIN Includes */
+#define ADC_MAX_VALUE 1023
+#define PWM_MIN_DUTY_CYCLE 0.05
+#define PWM_MAX_DUTY_CYCLE 0.10
+/* USER CODE END Includes */
+
+/* Private typedef -----------------------------------------------------------*/
+/* USER CODE BEGIN PTD */
+
+/* USER CODE END PTD */
+
+/* Private define ------------------------------------------------------------*/
+/* USER CODE BEGIN PD */
+/* USER CODE END PD */
+
+/* Private macro -------------------------------------------------------------*/
+/* USER CODE BEGIN PM */
+
+/* USER CODE END PM */
+
+/* Private variables ---------------------------------------------------------*/
+
+/* USER CODE BEGIN PV */
+
+/* USER CODE END PV */
+
+/* Private function prototypes -----------------------------------------------*/
+void SystemClock_Config(void);
+/* USER CODE BEGIN PFP */
+
+/* USER CODE END PFP */
+
+/* Private user code ---------------------------------------------------------*/
+/* USER CODE BEGIN 0 */
+
+/* USER CODE END 0 */
+
+/**
+  * @brief  The application entry point.
+  * @retval int
+  */
+int main(void)
+{
+  /* USER CODE BEGIN 1 */
+
+  /* USER CODE END 1 */
+
+  /* MCU Configuration--------------------------------------------------------*/
+
+  /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
+  HAL_Init();
+
+  /* USER CODE BEGIN Init */
+
+  /* USER CODE END Init */
+
+  /* Configure the system clock */
+  SystemClock_Config();
+
+  /* USER CODE BEGIN SysInit */
+
+  /* USER CODE END SysInit */
+
+  /* Initialize all configured peripherals */
+  MX_GPIO_Init();
+  MX_USART2_UART_Init();
+  MX_SPI1_Init();
+  MX_SPI2_Init();
+  MX_TIM1_Init();
+  /* USER CODE BEGIN 2 */
+
+
+  /* USER CODE END 2 */
+
+  /* Infinite loop */
+  /* USER CODE BEGIN WHILE */
+  while (1)
+  {
+	  /*
+	  HAL_StatusTypeDef  HAL_SPI_Receive(SPI_HandleTypeDef *hspi, uint8_t *pData, uint16_t Size, uint32_t Timeout);//receive data
+	  HAL_StatusTypeDef  HAL_SPI_Transmit(SPI_HandleTypeDef *hspi, uint8_t *pData, uint16_t Size, uint32_t Timeout);//send data
+	  uint32_t adcValue = HAL_ADC_GetValue(&hadc1); //Read the ADC Value
+	  uint32_t pwmValue = (adcValue * timerPeriod) / 4095; //Map the ADC Value to PWM Duty Cycle
+	  __HAL_TIM_SET_COMPARE(&htim2, TIM_CHANNEL_1, pwmValue);// Update the PWM Duty Cycle
+	  */
+	  HAL_GPIO_WritePin(GPIOB, GPIO_PIN_0, GPIO_PIN_RESET); // CS low
+	  uint8_t pRTxData[3];
+	  uint8_t txData[10] = {0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0A};
+	  HAL_SPI_TransmitReceive(SPI1, txData, pRTxData, 1, 1000);
+	  uint32_t pwm=0;
+	  HAL_GPIO_WritePin(GPIOB, GPIO_PIN_0, GPIO_PIN_SET); // CS high
+	  set_pwm_from_adc(pRTxData);

```

Comment:
so `set_pwm_from_adc` has a parameter of type `uint16_t` but you just passed in a pointer to an array. the only reason this compiles is because in c a pointer just boils down to a uint32_t which then got implicitly casted to a uint16_t. But this will definitely not work as you expect.

From the TX buffer, you need to extract the bytes to create the 10 bit adc value you are looking for. Refer to the datasheet in order to decode it properly..